### ISSUE TITLE
Auth.signedIn event triggered when not supposed to

### ISSUE THREAD

**Issue by chirastefan** [supabase-community/supabase-flutter#260]
**Description**

The `Auth.signed` event gets triggered even when accessing a verify email deep link from mail or when unlocking the device.

```
client?.auth.onAuthStateChange.listen((data) { ... })
```

**Expected behavior**
`Auth.signedIn` should be called only when you sign in with `signInWithPassword` or other signIn methods

**Environment**
 - Physical Device: Samsung Galaxy S7 Edge (Android 8.0.0)
 - OS: macOS 12.5.1 
 - Flutter 3.3.5
 - Dart 2.18.2 • DevTools 2.15.0

**Comment by petyots** [comment]
Are you by any chance using Getx?

**Comment by dshukertjr** [comment]
@chirastefan 
This is actually working as expected. There are multiple occasions where `AuthChangeEvent.signedIn` event is triggered, one of which is when email verification is complete.

Could I ask why you thought `AuthChangeEvent.signedIn` would be called only when signing in with `signInWithPassword`? Maybe we could provide better docs around it.

**Comment by chirastefan** [comment]
> 

hi, yes, I'm using Getx, why?

**Comment by chirastefan** [comment]
yes, I noticed that the `AuthChangeEvent.signedIn` event is triggered even for other events, for example, `auth.resetPasswordForEmail()`

Well what's the purpose of having `enum AuthChangeEvent {
  passwordRecovery,
  signedIn,
  signedOut,
  tokenRefreshed,
  userUpdated,
  userDeleted,
}` if only signedIn and signedOut are called? 

I would expect `AuthChangeEvent.passwordRecovery` to be called when I call `auth.resetPasswordForEmail()`, `AuthChangeEvent.userUpdated` when the email is verified, but instead only `AuthChangeEvent.signedIn` is called every time for the above events/actions.

If I call `auth.resetPasswordForEmail()` I get an access token and an `AuthChangeEvent.signedIn` event for that listener. 
I'm using that access token and that listener to check if the user is logged in or not in my app. So in this case, after calling `auth.resetPasswordForEmail()` I automatically log in. Is this the intended flow or am I doing something wrong?

The same case is for verifying the email.

**Comment by dshukertjr** [comment]
>I would expect AuthChangeEvent.passwordRecovery to be called when I call auth.resetPasswordForEmail(), AuthChangeEvent.userUpdated when the email is verified, but instead only AuthChangeEvent.signedIn is called every time for the above events/actions.

`AuthChangeEvent.passwordRecovery` is called when you call `auth.resetPasswordForEmail()`, click on the recovery link sent to the inbox and come back to the app via deep link so that you can then redirect the user to a password reset page to reset the password.

**Comment by chirastefan** [comment]
I understand that flow and I did that(as mentioned in the documentation), but as I've mentioned above, that event does not seem to trigger an `AuthChangeEvent. passwordRecovery` event, only a `AuthChangeEvent. signedIn`

**Comment by dshukertjr** [comment]
@chirastefan 
I cannot seem to reproduce your issue. I just tested it, and I can confirm that upon opening the password recovery link, I get `AuthChangeEvent.signedIn` followed by `AuthChangeEvent.passwordRecovery`. This is working as expected, as upon clicking email confirmation, you are signed into the app, hence the initial `signedIn` event, and then followed by the `passwordRecovery` event because the deeplink was for password recovery. 

You can see the code [here](https://github.com/supabase/gotrue-dart/blob/main/lib/src/gotrue_client.dart#L408) where the events are being sent. 

In your code, you can handle password recovery event like this in your code, and you should see it working.

```dart
supabase.auth.onAuthStateChange.listen((data) {
    final event = data.event;
    if(event== AuthChangeEvent.passwordRecovery) {
      // redirect user to update password page
    }
  });
}
```

>On what should I base the logic for logged in or not?

If `session` is not null in your `onAuthStateChange`, the user is signed in.

If you are still not seeing `AuthChangeEvent.passwordRecovery` event in your app, could you provide a detailed steps to reproduce, or perhaps create a loom video or something to share how you are testing it?

**Comment by chirastefan** [comment]
Hi, I managed to reproduce the same flow, `signedIn` event and then a `passwordRecovery` event. 
My point is that I cannot use that event listener or the `auth.currentSession` to check if the user is logged in, because that will be non-null even for an email confirmation, password recovery or other events.
So I cannot use that event listener to set for example a loggedIn bool flag on the client and I also cannot use the `currentSession` to check if the user is logged in or not.
Shouldn't the signedIn event be trigged only when the user signs in with auth.signInWithPassword() or other sign in methods? and shouldn't `auth.currentSession` be non-null only when the user logs in with `auth.signInWithPassword()` or other sign in methods?

**Comment by chirastefan** [comment]
"4. opens the email and clicks on the link, which takes the user back to the app <- At this point the user is logged in." Why? it's not necessary to be logged in in order to change your password.
"If a user is changing their password, they better be a user who has verified their identity, hence should be logged in". So the listener `auth.onAuthStateChange` and/or the `auth.currentSession` are not useable when it comes to checking if the user is logged in or not, because it's triggered(the event) and not null(currentSession) for other events(password reset, confirm email, etc)

**Comment by dshukertjr** [comment]
>Why? it's not necessary to be logged in in order to change your password.

That is just how Supabase works right now. Once you click on the password recovery link, you are logged in.

**Comment by chirastefan** [comment]
I guess this happens not only for the reset password feature but for the email confirmation also.

**Comment by dshukertjr** [comment]
@chirastefan Yup, both email confirmation and password reset case, once the user clicks on the email link and are brought back to the app, they are considered signed in. This is how Supabase works.

**Comment by talamaska** [comment]
> @chirastefan I cannot seem to reproduce your issue. I just tested it, and I can confirm that upon opening the password recovery link, I get `AuthChangeEvent.signedIn` followed by `AuthChangeEvent.passwordRecovery`. This is working as expected, as upon clicking email confirmation, you are signed into the app, hence the initial `signedIn` event, and then followed by the `passwordRecovery` event because the deeplink was for password recovery.
> 
> You can see the code [here](https://github.com/supabase/gotrue-dart/blob/main/lib/src/gotrue_client.dart#L408) where the events are being sent.
> 
> In your code, you can handle password recovery event like this in your code, and you should see it working.
> 
> ```dart
> supabase.auth.onAuthStateChange.listen((data) {
>     final event = data.event;
>     if(event== AuthChangeEvent.passwordRecovery) {
>       // redirect user to update password page
>     }
>   });
> }
> ```
> 
> > On what should I base the logic for logged in or not?
> 
> If `session` is not null in your `onAuthStateChange`, the user is signed in.
> 
> If you are still not seeing `AuthChangeEvent.passwordRecovery` event in your app, could you provide a detailed steps to reproduce, or perhaps create a loom video or something to share how you are testing it?

I know when this recovery event is not fired, when we chose AuthFlowType.pkce. I don't know if this expected. But it would be nice if we had the recovery password flow implement in all authFlowTypes.

**Comment by talamaska** [comment]
@dshukertjr I would love to hear some feedback about my comment. I know you might have a bunch of other stuff to do, but this is a critical issue,

**Comment by dshukertjr** [comment]
@talamaska 
Would you open a separate issue as it seems like a separate issue from the original issue of this thread. I'm going to close this issue as the original issue is the intended behavior of Supabase.
